Decentralized prover

AllFi opened the pull request:
The motivation behind this PR is to take a step forward towards enabling permissionless operators. To achieve this objective, the PR introduces the concept of a pending commitments queue.

In this PR, the following changes were implemented:

* The `pendingCommitments` have been added to the pool contract. It is a queue that stores all commitments waiting to be added to the Merkle tree.
* The `transact` function no longer updates merkle root and pool index or requires a tree proof and root after. Instead, it merely appends a new commitment to the `pendingCommitments`. This modification enables the submission of transactions in parallel since the ordering is now executed on-chain.
* The `appendDirectDeposits` function was also modified accordingly.
* A new method called `proveTreeUpdate` has been introduced. This method accepts a tree update proof and updates the Merkle root and pool index if the proof is valid. The operator who submits the proof receives a tree update fee.
* The fee structure has been divided into a transact fee and a tree update fee. The operator who submits a transaction receives the transact fee, while the operator who calls `proveTreeUpdate` receives the tree update fee. To ensure motivation for submitting a tree update proof, a new state variable called `minTreeUpdateFee` has been introduced.
* Additionally, a prover address has been added to the `transact` calldata to prevent operator frontrunning. 
Currently, there is no mechanism for the `appendDirectDeposit` function to guard against front-running attacks, making such attacks still possible in this case.
* To address the race condition during `proveTreeUpdate`, the operator submitting the first transaction can specify a privileged prover who will have a grace period to submit the tree update proof. To support this, a new field, gracePeriod, has been added.
* `AllowListOperatorManager` was implemented to support configuration with multiple operators.
* The memo message size has been added to simplify the implementation of the extra data later.
* Support for the new message prefix has been added. This is essential to support the new memo encryption scheme described in https://github.com/zkBob/libzeropool-zkbob/issues/14.

TODO:

- [x] Rename the `transact` function to allow the UI to distinguish between new and old calldata versions.
- [x] Add version in the `transact` calldata.
- [x] Remove forced exit?
- [x] A storage collision exists because the pool contract lacks a usable gap. Figure out how to fix it.
- [x] Add a gap for the future updates?
- [x] Update scripts.

Deploy:
- [x] Don't forget to call `setTokenSeller` after the upgrade

AllFi committed:
Move tree update in a separate function

AllFi committed:
Add grace period

AllFi committed:
Refactor code, add tests

r0wdy1 committed:
adds tests for dd

AllFi committed:
Fix message event, add comments

r0wdy1 committed:
adds tests for priority queue

AllFi committed:
Add tests, minor fixes

AllFi committed:
Add AllowListOperatorManager tests

AllFi committed:
Minor refactoring

AllFi committed:
Minor fixes

AllFi committed:
Refactor PriorityQueue

AllFi committed:
Fix formatting

AllFi committed:
Fix formatting

AllFi committed:
Fix formatting

AllFi committed:
Rename transact, add calldata version

AllFi committed:
Add commitment to RootUpdated event

AllFi committed:
Return gracePeriodEnd from pendingCommitment

AllFi committed:
Add memo message size

AllFi committed:
Add extra data to tests

AllFi committed:
Support new message prefix

AllFi committed:
Remove forced exit related code

k1rill-fedoseev commented on src/zkbob/ZkBobPool.sol:
```diff
@@ -61,18 +61,46 @@ abstract contract ZkBobPool is IZkBobPool, EIP1967Admin, Ownable, Parameters, Ex
 
     mapping(address => uint256) public accumulatedFee;
 
+    // TODO:
+    // Is it safe to just add deprecated gap here and
+    // call setTokenSeller() again after upgrade to fix storage layout?
+    // Additionally, it makes sense to add __gap as well.
+    // address private __deprecatedGap2;
```
Yes, this approach should work. Current token seller value will go to this slot and will stay unused, while we will re-initialize a new slot during the upgrade.

AllFi replied:
Good, thank you! I've added gaps in [3d39ff4](https://github.com/zkBob/zkbob-contracts/pull/97/commits/3d39ff4c1844f8d064ac073b1ee3801b5749bbb4).

k1rill-fedoseev commented on src/zkbob/manager/AllowListOperatorManager.sol:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    function _setOperators(
+        address[] memory _operators,
+        bool[] memory _allowed,
+        address[] memory _feeReceivers
```
I think we can keep `calldata` location here?

AllFi replied:
I've completely removed this function in https://github.com/zkBob/zkbob-contracts/pull/97/commits/ecc926b6341fcbefec73e72c9846b161f270934f. There is a bit of code duplication, but I believe it is still better than before.

k1rill-fedoseev commented on src/zkbob/manager/AllowListOperatorManager.sol:
```diff
@@ -0,0 +1,132 @@
+    function _setOperator(address _operator, bool _allowed, address _feeReceiver) internal nonZeroAddress(_operator) {
+        operators[_operator] = _allowed;
+        if (_feeReceiver != address(0) && _allowed) {
+            operatorFeeReceiver[_operator] = _feeReceiver;
```
When disabling operator, I think we should also reset `operatorFeeReceiver`.

AllFi replied:
In this case, the fee receiver address will not be able to withdraw fees from the pool contract after the operator has been disabled. But I am not sure about the significance of this feature. I suppose it may not be crucial.

k1rill-fedoseev replied:
You are right. Let's keep it then, so that fee receiver can still withdraw. But let's remove the `_feeReceiver != address(0)` condition then, so that fee receiver is always set when operator is enabled/re-enabled.

k1rill-fedoseev commented on src/zkbob/manager/AllowListOperatorManager.sol:
```diff
@@ -0,0 +1,132 @@
+    // if false, anyone can be an operator
+    bool public allowListEnabled;
+
+    // mapping of whitelisted operator addresses
+    mapping(address => bool) public operators;
+
+    // mapping of fee receivers for operators
+    mapping(address => address) public operatorFeeReceiver;
```
I think we can keep only one mapping here with fee receivers, and don't occupy more storage slots with bool flags.

AllFi replied:
Agree, I'll remove operators mapping if we decide that the issue described above isn't particularly significant.

AllFi replied:
I guess it is not possible if we want to allow fee receivers to claim fees after the operator has been disabled.

k1rill-fedoseev replied:
We can pack it in a single struct of `address` and `bool`, compiler will put them in a single storage slot in such case

AllFi replied:
Done in https://github.com/zkBob/zkbob-contracts/pull/97/commits/8d193842f3836649ddea99afce13e6d34225cfcb

k1rill-fedoseev commented on src/zkbob/utils/PriorityQueue.sol:
```diff
@@ -0,0 +1,81 @@
+    uint64 timestamp;
+}
+
+/// @author Matter Labs
```
Should we put a link or some specific license?

AllFi replied:
I've added the copyright notice in https://github.com/zkBob/zkbob-contracts/pull/97/commits/6d16f95b0ac5a7bbefdf566e022423d8e7062308. I guess this should be sufficient, considering that Matter Labs uses the MIT License.

k1rill-fedoseev commented on src/zkbob/utils/PriorityQueue.sol:
```diff
@@ -0,0 +1,81 @@
+    using PriorityQueue for Queue;
+
+    /// @notice Container that stores pending commitments
+    /// @param data The inner mapping that saves poending commitment by its index
```
```suggestion
    /// @param data The inner mapping that saves pending commitment by its index
```

k1rill-fedoseev commented on src/zkbob/utils/PriorityQueue.sol:
```diff
@@ -0,0 +1,81 @@
+    uint256 commitment;
+    address prover;
+    uint64 fee;
+    uint64 timestamp;
```
I think it's worth packing timestamp in `uint32`, so that all `prover`, `fee` and `timestamp` sit in a single storage slot.

AllFi replied:
Yes, thank you. I didn't think about it. Hope users from 2106 will not hate us :)

AllFi replied:
Fixed in https://github.com/zkBob/zkbob-contracts/pull/97/commits/6d16f95b0ac5a7bbefdf566e022423d8e7062308

k1rill-fedoseev commented on src/zkbob/utils/PriorityQueue.sol:
```diff
@@ -0,0 +1,81 @@
+/// @author Matter Labs
+/// @dev The library provides the API to interact with the priority queue container
+/// @dev Order of processing operations from queue - FIFO (Fist in - first out)
+library PriorityQueue {
```
Why is it called a priority queue if there is no priority? :)
Just a regular queue / fifo?

AllFi replied:
Renamed it in https://github.com/zkBob/zkbob-contracts/pull/97/commits/6d16f95b0ac5a7bbefdf566e022423d8e7062308.

k1rill-fedoseev commented on src/zkbob/ZkBobPool.sol:
```diff
@@ -108,6 +136,14 @@ abstract contract ZkBobPool is IZkBobPool, EIP1967Admin, Ownable, Parameters, Ex
+    /**
+     * @dev Throws if minimal tree update fee is not set.
+     */
+    modifier minTreeUpdateFeeIsSet() {
```
Is this really necessary, if we will set the min fee during the upgrade anyway?

AllFi replied:
Yes, indeed. It seems a bit excessive. Removed it in https://github.com/zkBob/zkbob-contracts/pull/97/commits/118287802305a9331f529cb19051b1e979220f9b.

k1rill-fedoseev commented on src/zkbob/ZkBobPool.sol:
```diff
@@ -512,4 +499,39 @@ abstract contract ZkBobPool is IZkBobPool, EIP1967Admin, Ownable, Parameters, Ex
+        // We calculate the beggining of the grace period either from the timestamp of the last tree update,
+        // or from the timestamp of the commitment, whichever is greater.
+        uint64 timestamp = commitmentTimestamp;
+        if (timestamp < lastTreeUpdateTimestamp) {
+            timestamp = lastTreeUpdateTimestamp;
+        }
```
So if I submit 100 transactions before someone's else's tx, their effective grace period can be `101 * gracePeriod`? Is this intended?

AllFi replied:
Yes, when you put it this way it doesn't sound good. My original thought was to ensure that each operator had a fair grace period even if other operators handled their commitments slowly or didn't do anything at all. So it seems we have these options:
* Keep the implementation as is. Users might experience some form of DOS, but I don't see any direct economic incentive to do so.
* Remove logic related to lastTreeUpdateTimestamp. Operators might not have a grace period, meaning other operators would have an incentive to send tree update proofs right before the end of the grace period to compete for the next proof.

There probably are some other options, but they might be more complicated. To be honest, I'm also not sure which approach is better.

k1rill-fedoseev replied:
I think timely proof inclusion should be prioritized over operators grace period profitability. 
In the current setting, if all 100 transactions in the queue are sent by different operators, it will become quite a bottleneck and take a lot of time to clear the whole queue of pending tree updates. However, if their grace periods would have expired sooner, single independent operator will be able to clear the whole queue at once more efficiently (even without zk proof batching).
Also, if some operator sees that they won't be able to fit their transaction inside the grace period and worry about potential race condition and transaction reverts, they may choose to not send tree proof at all.

AllFi replied:
Yes, it does make sense. We should prioritize liveness over fairness to the operators. Modified this logic in [d5648f7](https://github.com/zkBob/zkbob-contracts/pull/97/commits/d5648f73b2421878973388eea36374bdf7862453),

k1rill-fedoseev commented on src/zkbob/ZkBobPool.sol:
```diff
@@ -171,6 +227,27 @@ abstract contract ZkBobPool is IZkBobPool, EIP1967Admin, Ownable, Parameters, Ex
+     * Callable only by the contract owner / proxy admin.
+     * @param _minTreeUpdateFee new minimal fee.
+     */
+    function setMinTreeUpdateFee(uint64 _minTreeUpdateFee) external onlyOwner {
```
There is a censorship vector here, setting a too large min fee will prevent everyone from submitted transactions. Maybe we should make it immutable?

k1rill-fedoseev replied:
Well, immutable values still can be changed during contracts upgrade, which are usually put behind timelocks. Of course we can use the time lock for `owner` as well, not just the upgradeability owner. I think it's fine to keep it this way for the upcoming release. 

One other interesting idea I had was to allow only slow changes in min fee, e.g. we can allow only +-20% min fee change per 24h.

AllFi replied:
Okay, let's keep it as is for now. Yes, it is a good idea. I think we'll need to introduce more holistic logic to manage this value, especially if we implement batching.

AllFi committed:
Update src/zkbob/utils/PriorityQueue.sol

Co-authored-by: Kirill Fedoseev <kirill@blockscout.com>

AllFi committed:
Refactor AllowListOperatorManager

AllFi committed:
Rename PriorityQueue, add copyright, uint32 timestamp

AllFi committed:
Remove minTreeUpdateFeeIsSet modifier

AllFi committed:
Pack fee reciever and allowed in a single struct

AllFi committed:
Remove lastTreeUpdateTimestamp

AllFi committed:
Add gaps, fix testNativeWithdrawal

AllFi committed:
Fix formatting

k1rill-fedoseev reviewed (approved):
lgtm

EvgenKor commented on src/zkbob/ZkBobPool.sol:
```diff
@@ -24,18 +24,19 @@ import "../utils/Ownable.sol";
     bytes4 internal constant MESSAGE_PREFIX_COMMON_V1 = 0x00000000;
-    uint256 internal constant FORCED_EXIT_MIN_DELAY = 1 hours;
-    uint256 internal constant FORCED_EXIT_MAX_DELAY = 24 hours;
+    bytes4 internal constant MESSAGE_PREFIX_COMMON_V2 = 0x00000002;
```
Should we rename this and other corresponding constants in a more clear way? e.g. `MESSAGE_ENCRYPTION_ECDH` (0), `MESSAGE_ENCRYPTION_PLAIN` (1), `MESSAGE_ENCRYPTION_SYMMETRIC` (2)

AllFi replied:
Yeah, I've considered it. It appears to me that memo encryption falls outside the scope of contract logic. Additionally, `MESSAGE_PREFIX_COMMON_V1` could be removed once the system stabilizes. Therefore, from the contract perspective, there would be two possible prefixes: one for common transactions and one for dd transactions. That's why I prefer the current names. What do you think?

EvgenKor replied:
Of course I understand your logic. I'm afraid the naming differences between the library and contract code can confuse thirdparty. It's unclear the prefix is the same as an encryption scheme. Btw the memo encryption scheme is really outside of contract logic. But the scheme type is under that scope since we analyze it correctness here.

But it's a minor thing anyway and I don't insist here :)

AllFi replied:
Yes, it is a bit confusing indeed. I've tried to add some comments to clarify this, but all my attempts so far sound even more confusing than it currently is :) Let's keep it as is for now.

EvgenKor reviewed (approved):
lgtm

AllFi committed:
Update ZkBobPool.s.sol and Local.s.sol scripts

AllFi committed:
Add script to deploy new operator, modify upgrade script

AllFi committed:
Minor fixes

AllFi commented:
<details>
  <summary>Outdated comment</summary>
  I have updated the `ZkBobPool.s.sol`, `Local.s.sol`, and `DeployZkBobPoolModules.s.sol` scripts to accommodate modifications in this PR. Additionally, I have added `DeployAllowListOperatorManager.s.sol` to independently deploy `AllowListOperatorManager`, as it is not deployed within `DeployZkBobPoolModules.s.sol`. 
`DeployZkBobPoolModulesTest.t.sol` includes a fork test that tests the upgrade of the USDC pool on Optimism using the `DeployZkBobPoolModules` script.

Modifications in `ZkBobPool.s.sol` include:
* Deploying `AllowListOperatorManager` instead of `MutableOperatorManager`. The contract will be deployed with parameters `zkBobRelayer`, `zkBobRelayerFeeReceiver`, and `allowListEnabled` from `Env.s.sol`.
* Setting `accounting.kycProviderManager` if defined in `Env.s.sol`.
* Setting accounting limits to values identical to those in the `DeployZkBobPoolModules.s.sol` script.
* Setting `ZkBobPool.gracePeriod` and `ZkBobPool.minTreeUpdateFee`.
* Setting `AllowListOperatorManager.owner` to the owner specified in `Env.s.sol`, if defined

Modifications in `DeployZkBobPoolModules.s.sol` include:
* Setting `ZkBobPool.gracePeriod` and `ZkBobPool.minTreeUpdateFee`.
* Setting `ZkBobPool.tokenSeller`, as the previous value will be in `__deprecatedGap2` after the implementation update.
* Removing the increment of `txCount`

The `ZkBobPool.s.sol` might be used to deploy a new pool, a combination of `DeployZkBobPoolModules.s.sol` and `DeployAllowListOperatorManager.s.sol` might be used to upgrade the existing pool. 

*Notes:*
* The `DeployZkBobPoolModules.s.sol` script assumes that the proxyAdmin == pool.owner == deployer
* The parameters `gracePeriod` and `minTreeUpdateFee` should be updated before deployment
* `ZkBobPool.redeemer` is not configured by any of the mentioned scripts

<details>
<summary>Differences in storage layout between current pool implementation and migration/opt-BOB-to-USDC (USDC pool on Optimism)</summary>

```
OLD LAYOUT:
0) Ownable._owner
1) ZkBobAccounting.slot0
2) ZkBobAccounting.slot1
3) ZkBobAccoutning.tiers
4) ZkBobAccounting.snapshots
5) ZkBobAccounting.userStats
6) ZkBobPool.operatorManager
7) ZkBobPool.nullifiers
8) ZkBobPool.roots
9) ZkBobPool.all_messages_hash
10) ZkBobPool.accumulatedFee
11) ZkBobTokenSellerMixin.tokenSeller

NEW LAYOUT:
0) Ownable._owner                                  | SAME
1-3) ZkBobPool.__deprecatedGap                     | DEPRECATED
4) ZkBobPool.redeemer                              | REUSE
5) ZkBobPool.accounting AND ZkBobPool.pool_index   | REUSE
6) ZkBobPool.operatorManager                       | SAME
7) ZkBobPool.nullifiers                            | SAME
8) ZkBobPool.roots                                 | SAME
9) ZkBobPool.all_messages_hash                     | SAME
10) ZkBobPool.accumulatedFee                       | SAME
11) ZkBobPool.__deprecatedGap2                     | DEPRECATED
12-14) ZkBobPool.pendingCommitments                | NEW
15) ZkBobPool.gracePeriod AND minTreeUpdateFee     | NEW
16-65) ZkBobPool.__gap                             | NEW
66) ZkBobTokenSellerMixin.tokenSeller              | NEW
```
</details>
</details>

AllFi committed:
Add privileged prover in calldata

AllFi committed:
Update scripts

AllFi committed:
Add decentralization upgrade script

AllFi committed:
Modify migration script

AllFi committed:
Update scripts

AllFi committed:
Minor fixes

AllFi committed:
Fix format

AllFi committed:
Fix format

AllFi commented:
I believe this PR is ready for the final review.

**Deployment plan:**
1. Set `TOKEN_NUMERATOR` to 1000 before the upgrade.
2. Deploy the new pool implementation: `./script/deploy.sh ./script/scripts/NewZkBobPoolUSDCImpl.s.sol --tc DeployNewZkBobPoolUSDCImpl --broadcast --verify`
3. Set the proxies', provers', and operator manager owner addresses in the `DeployAllowListOperatorManager.s.sol` file.
4. Deploy the new operator manager: `./script/deploy.sh ./script/scripts/DeployAllowListOperatorManager.s.sol --tc DeployAllowListOperatorManager --broadcast --verify`
5. In the `MigrateDecentralization.s.sol` file, set the `newZkBobPoolImpl` and `newOperatorManager` addresses. Also, configure the `gracePeriod` and `minTreeUpdateFee` values appropriately.
6. Run the migration script: `./script/deploy.sh ./script/scripts/MigrateDecentralization.s.sol --tc MigrateDecentralization --sender 0xbe7D4E55D80fC3e67D80ebf988eB0E551cCA4eB7`
7. Use `run-latest.json` to prepare Safe proposal


**Parameter Values:**
* `gracePeriod`: I believe 3 minutes is a good starting point. It is short enough to maintain protocol liveness in a worst-case scenario, yet long enough to provide an adequate window in a normal scenario.
* `minTreeUpdateFee`: I computed the max, min, and mean transaction fees for the last 3 months for the OP-USDC pool. The results are as follows:
```
Mean: 0.16172255227554794 USD
Max: 20.095211035222057 USD
Min: 0.001998291024255525 USD
```
Considering these values, I think a fee of 0.1 USDC should cover the tree update cost in most cases. It's still relatively low, so the overall fee impact will remain minimal.
